File: sdk/es/src/stepauth/nameid.ts
/** Registered NameID type prefixes (`type:value`). */
export const NAMEID_EMAIL = "email";
export const NAMEID_PERSISTENT = "persistent";
export const NAMEID_GROUP = "group";
export const NAMEID_POLICY = "policy";

const REGISTERED = new Set([NAMEID_EMAIL, NAMEID_PERSISTENT, NAMEID_GROUP, NAMEID_POLICY]);

/** A parsed `type:value` identifier. */
export interface NameID {
  type: string;
  value: string;
}

/**
 * Split s on the FIRST colon only. The value may itself contain colons; only
 * the type prefix must be one of the registered types, and the value must be
 * non-empty. Throws on a malformed NameID.
 */
export function parseNameID(s: string): NameID {
  const i = s.indexOf(":");
  if (i < 0) throw new Error(`stepauth: nameid "${s}": missing ':'`);
  const type = s.slice(0, i);
  const value = s.slice(i + 1);
  if (!REGISTERED.has(type)) {
    throw new Error(`stepauth: nameid "${s}": unknown type "${type}"`);
  }
  if (value === "") throw new Error(`stepauth: nameid "${s}": empty value`);
  return { type, value };
}

/** Reports whether s is a well-formed NameID. */
export function isValidNameID(s: string): boolean {
  try {
    parseNameID(s);
    return true;
  } catch {
    return false;
  }
}

/** Reports whether n is a single person — the only kind allowed as principal.subject. */
export function isIndividual(n: NameID): boolean {
  return n.type === NAMEID_EMAIL || n.type === NAMEID_PERSISTENT;
}

export const email = (addr: string): string => `${NAMEID_EMAIL}:${addr}`;
export const persistent = (id: string): string => `${NAMEID_PERSISTENT}:${id}`;
export const groupId = (id: string): string => `${NAMEID_GROUP}:${id}`;
export const policyId = (id: string): string => `${NAMEID_POLICY}:${id}`;

File: sdk/go/stepauth/nameid.go
package stepauth

import (
	"fmt"
	"strings"
)

// NameID types. Email and Persistent are individuals.
const (
	NameIDEmail      = "email"
	NameIDPersistent = "persistent"
	NameIDGroup      = "group"
	NameIDPolicy     = "policy"
)

// NameID is a parsed typed identifier "type:value".
type NameID struct {
	Type  string
	Value string
}

// ParseNameID splits s on the FIRST colon; type must be registered and value non-empty.
func ParseNameID(s string) (NameID, error) {
	t, v, ok := strings.Cut(s, ":")
	if !ok {
		return NameID{}, fmt.Errorf("stepauth: nameid %q: missing ':'", s)
	}
	switch t {
	case NameIDEmail, NameIDPersistent, NameIDGroup, NameIDPolicy:
	default:
		return NameID{}, fmt.Errorf("stepauth: nameid %q: unregistered type %q", s, t)
	}
	if v == "" {
		return NameID{}, fmt.Errorf("stepauth: nameid %q: empty value", s)
	}
	return NameID{Type: t, Value: v}, nil
}

// String renders the NameID to wire form.
func (n NameID) String() string { return n.Type + ":" + n.Value }

// IsIndividual reports whether the NameID is a single person, the only kind
// allowed as principal.subject.
func (n NameID) IsIndividual() bool {
	return n.Type == NameIDEmail || n.Type == NameIDPersistent
}

// ValidNameID reports whether s parses as a well-formed NameID.
func ValidNameID(s string) bool {
	_, err := ParseNameID(s)
	return err == nil
}

// Email returns an email-typed NameID in wire form, e.g. Email("a@b.com") == "email:a@b.com".
func Email(addr string) string { return NameIDEmail + ":" + addr }

// Persistent returns a persistent-typed NameID in wire form.
func Persistent(id string) string { return NameIDPersistent + ":" + id }

// GroupID returns a group-typed NameID in wire form.
func GroupID(id string) string { return NameIDGroup + ":" + id }

// PolicyID returns a policy-typed NameID in wire form.
func PolicyID(id string) string { return NameIDPolicy + ":" + id }
